4810: Added integration tests project

adrian-fitsios opened the pull request:
- added sample data for integration tests
- implemented integration tests for 'get file names'

adrian-fitsios committed:
4810: Added integration tests project and implemented integration tests for 'get file names'

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Features/Get Solution Documents.3657.feature:
```diff
@@ -0,0 +1,38 @@
+﻿Feature:  Display uploaded file names
+    As BuyingCatalogueService
```
As *the* BuyingCatalogueService

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Features/Get Solution Documents.3657.feature:
```diff
@@ -0,0 +1,38 @@
+﻿Feature:  Display uploaded file names
+    As BuyingCatalogueService
+    I want make GET documents requests for a solution
```
I want *to* make GET documents requests for a solution

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Features/Get Solution Documents.3657.feature:
```diff
@@ -0,0 +1,38 @@
+﻿Feature:  Display uploaded file names
+    As BuyingCatalogueService
+    I want make GET documents requests for a solution
+    So that see which files does that soution have
```
So that I can see all available documents for a solution

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Steps/AzureBlobSteps.cs:
```diff
@@ -0,0 +1,35 @@
+namespace NHSD.BuyingCatalogue.Documents.API.IntegrationTests.Steps
+{
+    [Binding]
+    class AzureBlobSteps
```
Explicit `internal` access modifier

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Steps/AzureBlobSteps.cs:
```diff
@@ -0,0 +1,35 @@
+    [Binding]
+    class AzureBlobSteps
+    {
+        private AzureBlobStorageScenarioContext context;
```
New line after private field

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Steps/HttpClientSteps.cs:
```diff
@@ -0,0 +1,63 @@
+namespace NHSD.BuyingCatalogue.Documents.API.IntegrationTests.Steps
+{
+    [Binding]
+    class HttpClientSteps
```
Explicit `internal` access modifier

adrian-fitsios replied:
👍

bjss-chris replied:
This appears to be missing in your most recent commit.

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Support/AzureBlobStorageScenarioContext.cs:
```diff
@@ -0,0 +1,65 @@
+        private const string ConnectionString = @"UseDevelopmentStorage=true";
+        private const string ContainerName = "container-1";
+        private const string SampleDataPath = "SampleData";
+        public Dictionary<string, string> SolutionIdsToGuids = new Dictionary<string, string>();
```
To read-only auto property

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Support/AzureBlobStorageScenarioContext.cs:
```diff
@@ -0,0 +1,65 @@
+        public string TryToGetGuidFromSolutionId(string solutionId)
+        {
+            if (!SolutionIdsToGuids.TryGetValue(solutionId, out string slnId))
+            {
```
Is a `TryGetValue` necessary as you map solution IDs to a `Guid` in your **Given**?

If a solution ID doesn't have a corresponding `Guid` does it not imply that something has gone wrong with the test setup and therefore it might be better to allow a `KeyNotFoundException` to be thrown?

adrian-fitsios replied:
Good suggestion, however it is necessary as TryGetValue allows for introducing test cases with invalid solution (solution that doesn't exist). The scenario #1, in its last test case, asserts that when we query for 'sol99' it returns an empty list. Should we remove this and let it throw, it would always throw and never hit the endpoint.

bjss-chris replied:
Ah, I missed that. Thanks for clarifying.

When a solution ID doesn't exist might it make sense to return `Guid.Empty.ToString()` so that test cases such as this follow the same convention (i.e. using a `Guid` in place of the solution ID) as the test cases where the solution ID does exist?

Also, is there any way of improving the variable names to help distinguish between `solutionId` and `slnId` without resorting to Hungarian-style naming?

adrian-fitsios replied:
Good idea! Resorted to solutionIdAsGuid for now. Hope that doesn't make it too confusing.

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Support/IntegrationSetUpFixture.cs:
```diff
@@ -0,0 +1,24 @@
+        [Binding]
+        public class IntegrationSetupFixture
+        {
+            private readonly AzureBlobStorageScenarioContext _scenarioContext;
```
New line after the private field and before the constructor

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Support/IntegrationSetUpFixture.cs:
```diff
@@ -0,0 +1,24 @@
+                this._scenarioContext = scenarioContext;
+            }
+
+            [AfterScenario()]
```
You can omit the brackets from the attribute.

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Features/Get Solution Documents.3657.feature:
```diff
@@ -0,0 +1,38 @@
+		| sol4       | File                                                                               |
+		| sol5       | File with whitespace in its name                                                   |
+		| sol6       | děkuji.pdf,Благодаря ти.blin,ありがとうございました.txt,תודה.jpeg, شكرا لكم.png, 谢 |
+		| sol99      |                                                                                    |
```
The vertical bars would be better if aligned but understand that it's not possible as the non-ASCII characters are not fixed-width.

bjss-chris reviewed (commented):
Looks good - just a few small suggestions with code style.

We also probably want to remove .editorconfig from the .dockerignore so that changes in the default warning severity don't cause the docker build to fail.

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Support/IntegrationSetUpFixture.cs:
```diff
@@ -0,0 +1,24 @@
+            private readonly AzureBlobStorageScenarioContext _scenarioContext;
+            public IntegrationSetupFixture(AzureBlobStorageScenarioContext scenarioContext)
+            {
+                this._scenarioContext = scenarioContext;
```
You can omit the `this.`

adrian-fitsios replied:
👍

adrian-fitsios committed:
4810: PR Comments

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Steps/AzureBlobSteps.cs:
```diff
@@ -6,9 +6,10 @@
-    class AzureBlobSteps
+    internal class AzureBlobSteps
     {
         private AzureBlobStorageScenarioContext context;
```
This field can be made `readonly` and while I don't like underscores in field names it probably ought to have one in this case to remain consistent with the rest of the project.

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Steps/AzureBlobSteps.cs:
```diff
@@ -6,9 +6,10 @@
+
         public AzureBlobSteps(AzureBlobStorageScenarioContext context)
         {
             this.context = context;
```
`this` can be omitted if adding an underscore to the `context` field.

adrian-fitsios replied:
👍

adrian-fitsios committed:
4810: PR Comments

- Changed few things as per PR Comments
- Made the tests actually run in parallel
- Created another API which represents the API in a broken state
- Added tests for when the blob storage is not available

bjss-chris commented on docker-compose.yml:
```diff
@@ -12,13 +12,25 @@ services:
     ports:
       - "8090:80"
+
+  nhsd.buyingcatalogue.document.api.broken:
```
Although 'broken' is in the name I wonder if it might be worth including a comment to explain that the omission of `networks` is deliberate so that communication with the storage emulator fails?

(As the correct environment variables are included it might help to make it absolutely clear that it's not a mistake.)

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Steps/HttpClientSteps.cs:
```diff
@@ -16,21 +16,27 @@ class HttpClientSteps
         public HttpClientSteps(ScenarioContext context, AzureBlobStorageScenarioContext azureBlobStorageScenarioContext)
         {
             _context = context;
+            _context["RootUrl"] = "http://localhost:8090/api/v1/Solutions";
```
Might it be worth refactoring the URL to a constant so that it is immediately clear to anyone looking at the code that this is the URL for the working service?

For example

```c#
const string workingServiceUrl = "http://localhost:8090/api/v1/Solutions";
_context["RootUrl"] = workingServiceUrl;
```

This would just mean anyone looking at the code doesn't have check elsewhere in the code or the docker-compose YAML file to confirm which URL corresponds to which version of the service.

The constant is local in the example above but could be class level if you think that is a better fit. The name above is also just a suggestion.

Another approach could be a nested class, for example:

```c#
private static class ServiceUrl
{
    internal const string Broken = "...";
    internal const string Working = "...";
}
```

If you think this is worthwhile please feel free to use whichever approach (and formatting) you prefer.

adrian-fitsios replied:
I like this approach, it's going to make it more maintainable.

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Steps/HttpClientSteps.cs:
```diff
@@ -16,21 +16,27 @@ class HttpClientSteps
+        [Given(@"the blob storage service is down")]
+        public void GivenTheBlobStorageServiceIsDown()
+        {
+            _context["RootUrl"] = "http://localhost:8091/api/v1/Solutions";
```
Use of constant for the URL as suggested for the constructor?

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Features/Get Solution Documents Service Not Available.3657.feature:
```diff
@@ -0,0 +1,15 @@
+﻿Feature:  Try to display file names when service is not available
```
'Try to display file names when *the storage* service is not available'?

adrian-fitsios replied:
👍

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Support/IntegrationSetUpFixture.cs:
```diff
@@ -1,6 +1,9 @@
+using NUnit.Framework;
 using TechTalk.SpecFlow;
 
+[assembly: Parallelizable(ParallelScope.Fixtures)]
```
Might it be worth creating an `AssemblyInfo.cs` file and moving the assembly level attribute to that file? (Most developers will be familiar with the `AssemblyInfo.cs` convention from the .NET Framework.)

(We can't use the .csproj `AssemblyAttribute` element because it only supports `string` parameters.)

Other than `InternalsVisibleTo`, are there any assembly level attributes used in BAPI? If so, what is the convention there?

adrian-fitsios replied:
I'm in on the idea of creating Assembly Info property file as it is the cleanest way.

Not that I am aware of, I think BAPI was 'lucky' enough to use only string params and keeps them in the project file.

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Features/Get Solution Documents.3657.feature:
```diff
@@ -15,7 +15,7 @@ Background:
 		| sol6       | děkuji.pdf                       |
 		| sol6       | Благодаря ти.blin                |
-		| sol6       | ありがとうございました.txt         |
+		| sol6       | ありがとうございました.txt                  |
```
Vertical bar closer in line to the others? (Aware it can't be perfect due to non-fixed-width characters).

adrian-fitsios replied:
👍

adrian-fitsios committed:
4810: Feature file refactor

Renamed the feature & removed unnecessary set-up step

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Features/Get Solution Documents.3657.feature:
```diff
@@ -28,11 +28,11 @@ Scenario Outline: 1. Correct file names are returned for a solution
+		| sol5       | File with whitespace in its name                                        |
 		| sol6       | děkuji.pdf,Благодаря ти.blin,ありがとうございました.txt,תודה.jpeg, شكرا لكم.png, 谢 |
-		| sol99      |                                                                                    |
+		| sol99      |                                                                         |
```
I think the vertical bars might have looked better where they were (closer in line to the bar for sol6).

adrian-fitsios replied:
👍

bjss-chris reviewed (commented):
I've made a few suggestions but looks good overall.

adrian-fitsios committed:
4810: PR Comments

bjss-chris commented on test/NHSD.BuyingCatalogue.Documents.API.Integration.Tests/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1 @@
+﻿[assembly: NUnit.Framework.Parallelizable(NUnit.Framework.ParallelScope.Fixtures)]
```
You could include a `using NUnit.Framework;` and then this line could be simplified to `[assembly: Parallelizable(ParallelScope.Fixtures)]`.

This is fine for now though.

adrian-fitsios committed:
4810: Cleaner way of setting-up files

adrian-fitsios committed:
4810: More comments

bjss-chris reviewed (approved):
Good work @assoomro. This gives us an excellent basis on which to build additional integration tests and automated tests as well as an easy means of spinning up the DAPI when working on other areas of the application.
